pkg/operator/agent_manager.go

```diff
@@ -1,14 +1,221 @@
 package operator
 
-import "k8s.io/client-go/pkg/api/v1"
+import (
+	"fmt"
 
-var managedByOperatorLabel = map[string]string{
-	"managed-by": "container-linux-update-operator",
-}
+	"github.com/blang/semver"
+	"github.com/coreos/container-linux-update-operator/pkg/constants"
+	"github.com/coreos/container-linux-update-operator/pkg/version"
+	"github.com/golang/glog"
+
+	"k8s.io/client-go/pkg/api/unversioned"
+	"k8s.io/client-go/pkg/api/v1"
+	"k8s.io/client-go/pkg/apis/extensions/v1beta1"
+	"k8s.io/client-go/pkg/labels"
+)
+
+var (
+	managedByOperatorLabels = map[string]string{
+		"managed-by": "container-linux-update-operator",
+		"app":        agentDefaultAppName,
+	}
+
+	daemonsetName = "container-linux-update-agent-ds"
+)
 
 // updateAgent updates the agent on nodes if necessary.
-// It returns a value indicating whethre the update-agent is currently up to
-// date.
-func (k *Kontroller) updateAgent(nodeList *v1.NodeList) (bool, error) {
-	agentDaemonset, err := k.kc.DaemonSets(k.namespace).List(&v1.ListOptions{})
+//
+// NOTE: the version for the agent is assumed to match the versioning scheme
+// for the operator, thus our version is used to figure out the appropriate
+// agent version.
+// Furthermore, it's assumed that all future agent versions will be backwards
+// compatible, so if the agent's version is greater than ours, it's okay.
+func (k *Kontroller) runDaemonsetUpdate(agentImageRepo string) error {
+	agentDaemonsets, err := k.kc.DaemonSets(k.namespace).List(v1.ListOptions{
+		LabelSelector: labels.SelectorFromSet(labels.Set(managedByOperatorLabels)).String(),
+	})
+	if err != nil {
+		return err
+	}
+
+	if len(agentDaemonsets.Items) == 0 {
+		// No daemonset, create it
+		runErr := k.createAgentDamonset(agentImageRepo)
+		if runErr != nil {
+			return runErr
+		}
+		// runAgent succeeded, all should be well and converging now
+		return nil
+	}
+
+	// There should only be one daemonset since we use a well-known name and
+	// patch it each time rather than creating new ones.
+	if len(agentDaemonsets.Items) > 1 {
+		glog.Errorf("only expected one daemonset managed by operator; found %+v", agentDaemonsets.Items)
+		return fmt.Errorf("only expected one daemonset managed by operator; found %v", len(agentDaemonsets.Items))
+	}
+
+	agentDS := agentDaemonsets.Items[0]
+
+	var dsSemver semver.Version
+	if dsVersion, ok := agentDS.Annotations[constants.AgentVersion]; ok {
+		ver, err := semver.Parse(dsVersion)
+		if err != nil {
+			return fmt.Errorf("agent daemonset had version annotation, but it was not valid semver: %v[%v] = %v", agentDS.Name, constants.AgentVersion, dsVersion)
+		}
+		dsSemver = ver
+	} else {
+		glog.Errorf("managed daemonset did not have a version annotation: %+v", agentDS)
+		return fmt.Errorf("managed daemonset did not have a version annotation")
+	}
+
+	if dsSemver.LT(version.Semver) {
+		// daemonset is too old, update it
+		// TODO: perform a proper rolling update rather than delete-then-recreate
```

Comment:
Once https://github.com/coreos-inc/operator-client/pull/41 lands we can use that to do a proper upstream rolling update. Even without that patch, older versions of operator-client working on a < 1.6 cluster will do a manual rolling update.